File: bsp/adafruit-feather-m0-wifi/csharp/FeatherM0Wifi.cs
// Lamella.Boards.Adafruit.FeatherM0Wifi -- the Adafruit Feather M0 WiFi (ATSAMD21G18A, Cortex-M0+,
using System.Device.Gpio;
using Lamella.Generated;
using Lamella.Hardware;

namespace Lamella.Boards.Adafruit
{
    public sealed class FeatherM0Wifi
    {
        /// <summary>The wire identity this board advertises (lamella_wire::board_model).</summary>
        public static readonly int BoardModel = FeatherM0WifiBindings.BOARD_MODEL;

        /// <summary>The red LED beside the USB jack (Arduino D13) -- the board's blink target, and
        /// the only indicator visible without driving a protocol.</summary>
        public static readonly int LedPin =
            LogicalPin(FeatherM0WifiBindings.LED_PORT_BASE, FeatherM0WifiBindings.LED_PIN);

        /// <summary>The radio's chip-select line on the SPI bus it shares.</summary>
        public static readonly int WincChipSelectPin =
            LogicalPin(FeatherM0WifiBindings.WINC_SPI_CS_PORT_BASE, FeatherM0WifiBindings.WINC_SPI_CS_PIN);

        /// <summary>The radio's reset line, ACTIVE LOW: driving it <see cref="PinValue.Low"/> holds
        /// the WINC in reset.</summary>
        public static readonly int WincResetPin =
            LogicalPin(FeatherM0WifiBindings.WINC_RESET_N_PORT_BASE, FeatherM0WifiBindings.WINC_RESET_N_PIN);

        /// <summary>The radio's enable line, ACTIVE HIGH: driving it <see cref="PinValue.High"/>
        /// powers the WINC on. The opposite polarity from <see cref="WincResetPin"/>, which is a
        /// board fact rather than a convention and is why both are stated.</summary>
        public static readonly int WincChipEnablePin =
            LogicalPin(FeatherM0WifiBindings.WINC_CHIP_EN_PORT_BASE, FeatherM0WifiBindings.WINC_CHIP_EN_PIN);

        /// <summary>Binds this board's GPIO block to the driver table, so a program writes plain
        /// dotnet/iot -- <c>new GpioController()</c> -- and never names a Lamella type. Touching
        /// <see cref="FeatherM0Wifi"/> at all is what arms it, which is why a program constructs the
        /// board first.</summary>
        /// <remarks>A TYPE INITIALIZER rather than the instance constructor, for the reason
        /// <see cref="Lamella.Hardware.Buses.BindGpio"/> documents: the table refuses a second bind
        /// of the same kind rather than replacing it, and this class is instantiable and routinely
        /// constructed as a temporary. The language runs a type initializer once per program, so
        /// idempotence costs nothing and the table keeps its throw as a genuine-error detector.
        /// The bound value is a FACTORY, not a driver, so a program that never touches GPIO never
        /// constructs one.</remarks>
        static FeatherM0Wifi()
        {
            Buses.BindGpio(new GpioDriverFactory(MakeGpio));
        }

        private static GpioDriver MakeGpio() { return new Samd21GpioDriver(); }

        /// <summary>The family PORT driver this board bound, over every pin on the part.</summary>
        /// <remarks>THE SAME INSTANCE <see cref="GpioController"/> drives. One block has one
        /// driver, and handing out a second one over the same registers reads as working while the
        /// facade talks to the first -- see <see cref="Lamella.Hardware.Buses.ResolveSpi"/> for the
        /// full argument.</remarks>
        public GpioDriver CreateGpioDriver()
        {
            return Buses.ResolveGpio();
        }

        /// <summary>A GPIO controller over the SAMD21 PORT block.</summary>
        public GpioController CreateGpioController()
        {
            return new GpioController();
        }

        /// <summary>The driver's logical numbering: group * 32 + pin, with the group taken from the
        /// binding's port base. Kept here rather than in the driver because it converts a BOARD fact
        /// (which port a device sits on) into the family's pin space.</summary>
        static int LogicalPin(uint portBase, uint pin)
        {
            int group = portBase == Samd21Instances.PORTB_BASE ? 1 : 0;
            return group * 32 + (int)pin;
        }
    }
}

File: bsp/adafruit-feather-m0-wifi/swift/FeatherM0Wifi.swift
// HAND-WRITTEN thin board class for the Adafruit Feather M0 WiFi in Swift: turns the GENERATED
// AdafruitFeatherM0WifiBindings constants into pin descriptors. Nothing here is a hardware literal --
// every address, mask and polarity comes from the generated bindings, which come from
// bsp/adafruit-feather-m0-wifi/board.toml.

/// The Adafruit Feather M0 WiFi, as its indicator and its WINC1500's control lines.
///
/// The three WINC lines are ordinary GPIO on this board -- the SPI carries the data, but reset,
/// chip-enable and the interrupt are levels the host drives or reads. Their POLARITY is board
/// truth and is carried here rather than assumed: RESET_N and IRQN are active low, CHIP_EN is
/// active high, so `assert(wincReset)` holds the part in reset whichever way the line is wired.
public enum FeatherM0Wifi {
    /// The red LED beside the USB jack.
    public static var led: Samd21PinBinding {
        Samd21PinBinding(portBase: AdafruitFeatherM0WifiBindings.LED_PORT_BASE,
                         mask: AdafruitFeatherM0WifiBindings.LED_MASK,
                         activeLow: AdafruitFeatherM0WifiBindings.LED_ACTIVE_LOW == 1)
    }

    /// The WINC1500's reset line -- asserted (low) holds the part in reset.
    public static var wincReset: Samd21PinBinding {
        Samd21PinBinding(portBase: AdafruitFeatherM0WifiBindings.WINC_RESET_N_PORT_BASE,
                         mask: AdafruitFeatherM0WifiBindings.WINC_RESET_N_MASK,
                         activeLow: AdafruitFeatherM0WifiBindings.WINC_RESET_N_ACTIVE_LOW == 1)
    }

    /// The WINC1500's chip enable -- asserted (high) powers the part up.
    public static var wincChipEnable: Samd21PinBinding {
        Samd21PinBinding(portBase: AdafruitFeatherM0WifiBindings.WINC_CHIP_EN_PORT_BASE,
                         mask: AdafruitFeatherM0WifiBindings.WINC_CHIP_EN_MASK,
                         activeLow: AdafruitFeatherM0WifiBindings.WINC_CHIP_EN_ACTIVE_LOW == 1)
    }

    /// The WINC1500's interrupt line -- an INPUT: the module asserts it (low) when a message
    /// waits. Carried as the same descriptor type; only the direction a driver configures differs.
    public static var wincInterrupt: Samd21PinBinding {
        Samd21PinBinding(portBase: AdafruitFeatherM0WifiBindings.WINC_IRQN_PORT_BASE,
                         mask: AdafruitFeatherM0WifiBindings.WINC_IRQN_MASK,
                         activeLow: AdafruitFeatherM0WifiBindings.WINC_IRQN_ACTIVE_LOW == 1)
    }

    /// The WINC1500's chip select -- a SOFT select: the SERCOM does not drive it, the driver
    /// does, which is what lets one bus serve more than one device. Active low is the SPI
    /// convention rather than a board fact, so unlike the lines above it is stated here and not
    /// read from a generated `_ACTIVE_LOW` (the spi binding emits none).
    public static var wincChipSelect: Samd21PinBinding {
        Samd21PinBinding(portBase: AdafruitFeatherM0WifiBindings.WINC_SPI_CS_PORT_BASE,
                         mask: AdafruitFeatherM0WifiBindings.WINC_SPI_CS_MASK,
                         activeLow: true)
    }

    /// The SERCOM4 SPI wiring that reaches the WINC1500.
    ///
    /// `baudDivisor` and the clock mode are the CALLER's operating point, not board truth: the
    /// synchronous divisor is f_ref / (2 * f_sck) - 1, and f_ref is whatever generator the
    /// driver routes. SPI mode 0 (CPOL 0, CPHA 0) is what the WINC1500 expects.
    public static func wincSpi(baudDivisor: UInt32) -> Samd21SercomSpiBinding {
        Samd21SercomSpiBinding(
            sercomBase: AdafruitFeatherM0WifiBindings.WINC_SPI_SERCOM_BASE,
            gclkCoreId: AdafruitFeatherM0WifiBindings.WINC_SPI_GCLK_CORE_ID,
            apbcMask: AdafruitFeatherM0WifiBindings.WINC_SPI_APBC_MASK,
            pmuxFunc: AdafruitFeatherM0WifiBindings.WINC_SPI_PMUX_FUNC,
            pmuxMosiReg: AdafruitFeatherM0WifiBindings.WINC_SPI_PMUX_MOSI_REG,
            pmuxMosiShift: AdafruitFeatherM0WifiBindings.WINC_SPI_PMUX_MOSI_SHIFT,
            pincfgMosiReg: AdafruitFeatherM0WifiBindings.WINC_SPI_PINCFG_MOSI_REG,
            pmuxSckReg: AdafruitFeatherM0WifiBindings.WINC_SPI_PMUX_SCK_REG,
            pmuxSckShift: AdafruitFeatherM0WifiBindings.WINC_SPI_PMUX_SCK_SHIFT,
            pincfgSckReg: AdafruitFeatherM0WifiBindings.WINC_SPI_PINCFG_SCK_REG,
            pmuxMisoReg: AdafruitFeatherM0WifiBindings.WINC_SPI_PMUX_MISO_REG,
            pmuxMisoShift: AdafruitFeatherM0WifiBindings.WINC_SPI_PMUX_MISO_SHIFT,
            pincfgMisoReg: AdafruitFeatherM0WifiBindings.WINC_SPI_PINCFG_MISO_REG,
            dopo: AdafruitFeatherM0WifiBindings.WINC_SPI_DOPO,
            dipo: AdafruitFeatherM0WifiBindings.WINC_SPI_DIPO,
            baudDivisor: baudDivisor,
            cpol: 0,
            cpha: 0)
    }
}
